### ISSUE TITLE
Create  initial models for HTTP request

### ISSUE THREAD

**Issue by ehoner** [scalaz/scalaz-http#4]
These models are case class that represent the valid components of an HTTP request. There should be a model for each distinct type defined by the protocol. Reference RFCs:
* [RFC 7230](https://tools.ietf.org/html/rfc7230), HTTP/1.1: Message Syntax and Routing
* [RFC 7231](https://tools.ietf.org/html/rfc7231), HTTP/1.1: Semantics and Content
* [RFC 7232](https://tools.ietf.org/html/rfc7232), HTTP/1.1: Conditional Requests
* [RFC 7233](https://tools.ietf.org/html/rfc7233), HTTP/1.1: Range Requests
* [RFC 7234](https://tools.ietf.org/html/rfc7234), HTTP/1.1: Caching
* [RFC 7235](https://tools.ietf.org/html/rfc7235), HTTP/1.1: Authentication

The initial implementation does not need to cover all the variations, but needs to be sufficient that we can begin exploring the problem space. Prefer implementing more distinct cases rather than multiple variations of the same type. For example: 

Varied HTTP request headers (each element has a distinct type)
```scala
final case class Cookie(elements: List[String])

final case class Accept(value: List[`Media-Encoding`])

final case class Referer(value: URI)
```

Similar HTTP request headers (the headers values are the same type)
```scala
final case class Allow(values: List[Method])

final case class `Access-Control-Allow-Methods`(values: List[Method])
```

Also, prefer more coverage of the protocol than any single component. Only document what is partial for others to continue.

The names that I have chose are arbitrary, but they should reflect the protocol. There will be some overlap between request and response models, once the initial work has been done, managing the overlap can begin.

**Comment by syves** [comment]
Hi, 
I'm  a bit confused. Comments in Mobilize seem to indicate that we are building on top of http4s, if so then why do we need to model REQUEST and RESPONSE?

**Comment by sergei-shabanau** [comment]
@syves 
I don't think that building on top of http4s is what's decided. http4s is a competitor, and currently is based on cats. Unless there is a decision to use older versions that where built on scalaz 7...

**Comment by syves** [comment]
Hi Sergei, somehow I got that impression from some comments I read in Mobilize. Thank you for clarifying. The last version of http4s I used was based on Scalaz 7. Building on top of that seems wrong.

**Comment by syves** [comment]
@sergei-shabanau I'll continue reading up on redeyes and servant. And then dig into this issue.
